File: Problem/3307.cpp
#include <iostream>
#include <vector>

int acc = 0;
static const long long i = 1;
char kthCharacter(long long k, std::vector<int> &operations)
{
    if (k <= 1)
    {
        return (char)((int)'a' + acc % 26);
    }
    int jump = 0;
    // ceil log2k
    while ((i << jump) < k)
    {
        ++jump;
    }
    acc += operations[jump - 1];
    return kthCharacter(k - (i << (jump - 1)), operations);
}

File: Problem/3307.py
def kthCharacter(k, operations):
        
    def rec(val, acc):
        jump = 0
        while (1 << jump) < val:
            jump+=1
        if acc > 26:
            acc %= 26
        if val <= 1:
            return chr(ord('a') + acc)
        return rec(val - (1 << (jump-1)), acc + operations[jump-1])
    
    return rec(k, 0)

print(kthCharacter(10, [0,1,0,1]))